lib/postprocessing/image.js

```diff
@@ -192,7 +199,47 @@ async function read(infile) {
     return gm(`${infile}[0]`);
 }
 
-async function applyOperations(img, instructions) {
+function getImageSize(metadata) {
+    // The orientation corresponds to the TIFF orientation tag which ranges fromn 1 to 8
+    // Values 5 to 8 require a swap of width and height
+    // See https://www.awaresystems.be/imaging/tiff/tifftags/orientation.html or
+    // https://www.adobe.io/content/dam/udp/en/open/standards/tiff/TIFF6.pdf
+    if(CHANGE_ORIENTATION_LIST.includes(metadata.orientation)) {
+        return new Dimensions(metadata.pageGeometry.height, metadata.pageGeometry.width);
+    } else {
+        return new Dimensions(metadata.pageGeometry);
+    }
+}
+async function applyWatemark(img, rendition, directories) {
+    const instructions = rendition.instructions;
+    const targetBoundingBox = new Dimensions(instructions.width, instructions.height);
+
+    // TODO drop old watermark.widthPercent once no longer used by AEM
+    const watermarkScale = instructions.watermark.scale || (instructions.watermark.widthPercent / 100.0) || 1.0;
+
+    // determine image dimensions
+    const imageSize = getImageSize(rendition.intermediateMetadata);
+
+    // TODO drop old watermark.watermarkContent once no longer used by AEM
+    const watermarkSource = instructions.watermark.image || instructions.watermark.watermarkContent;
+    // download watermark
+    if(!watermarkSource) {
+        throw new GenericError(`post-processing: Missing watermark.image in instructions : ${JSON.stringify(rendition.instructions)}`);
+    }
+    const asset = await getAsset(watermarkSource, directories.in, WATERMARK_NAME);
+    // TODO: infuture should we enforce that watermark is a png ?
+    // TODO: infuture handle same watermark png across multiple renditions
+    const watermarkSize = new Dimensions(await gm(asset.path).size());
+
+    // calculate target sizes
+    const targetSize = imageSize.fitBoundingBox(targetBoundingBox, true);
+    const watermarkTargetSize = watermarkSize.fitBoundingBox(targetSize).scale(watermarkScale);
+
+    img.draw([`gravity Center image Over 0,0 ${watermarkTargetSize.width},${watermarkTargetSize.height} "${asset.path}"`]);
```

Comment:
I would turn `Center` in a constant so in the future it may be easier to support other placements.

Reply:
Yeah, we don't really know what future options look like. The other placement options from imagemagick (mostly in the corners) actually don't make much sense for watermarks. The other discussed option was to tile the watermark multiple times across the image, which will need some different logic anyway.